### DISCUSSION TITLE
Need advice on simple way to print

### DISCUSSION THREAD

**Discussion by AnalogWorld** [SpenceKonde/megaTinyCore#949] [Q&A] [answered]
I am new to Arduino and experimented a bit with Attiny85. I am now using Atttiny814 for more pins/memory migration.
I have received Tindie ATtiny1614/3224/etc breakout pcb  and flashed a blink and PWM program via Pickit 4 (Great board will make good stocking stuffer !)
I now would like to go to ADC experiments however I need to see what bit results are coming in from a analog input. What is the most simple way to accomplish printing data/file  to some GUI for debugging.? I am trying to get PYMCUPROG to work but I am having some path issues in Python.  I thought of writing the conversion result to EEprom and reading it with AVRDUDE (well actually AVRDUDESS as it is more graphically friendly).
I guess I really would like to stick with Pickit 4 as many protocols are listed adaptable to Pics & 0 series.   Many thanks to all for this forum, Megatinycore, and all the hard work that went into the program.  Any help appreciated however I am a bit of a newbie in the world of code.

Kind Regards
Art

**Comment by grandaspanna** [comment]
Probably the simplest is to grab a second USB serial adapter and connect to the TX/RX pins on the microcontroller. You can the use the standard Serial class to print whatever data you want and see that in a terminal window on your computer.

**Reply by SpenceKonde** [reply, marked as answer]
just throw a serial adapter onto the 6-pin FTDI header on the edge of the board. Standard pinout, pins as marked. The one catch is there's no autoreset because you can get autoreset without disabling UPDI unless you have a part with a PB4 (20 pins or 24 pins) as you can set it to use PB4 as an alternate reset pin and do the usual autoreset. I recommend not using the Arduino IDE's console window (it's terrible, lacks literally every feature on my serial console checklist); instead, have an external program (I highly recommend hTerm - it lets you view or send data as ascii, hex, binary, or decimal bytes (eg, to send a space you can send a ascii ' ', hex 20, decimal 32, or binary 00100000 and you can even mix and match those in a blob of data that you send when you hit enter, and for viewing received data, it's not either or - it's a checkbox for each so you could see the ascii characters and hex values (with different color background so you know which is which), and it lets you see the 4 modem control lines and gives full control over DTR and RTS. Lets you copy data sent or received in any of those 4 formats. It was written by an Arduino person so it's got the kind of things we need, you know?), and leave that open and connected while uploading, and remember to put something like `Serial.println("Start");` right after configuring the UART so you know when you're looking at new output.

**Comment by SpenceKonde** [comment]
Oh, and I'd be remiss if i didn't note that I'll be launching a dual port serial adapter probably this month that will have SerialUPDI programming on one port, and the other one as a normal serial port (also a serial adapter that has a switch to switch between Serial and UPDI modes)

**Reply by SpenceKonde** [reply]
Eurgh, nope. Board needs a respin.

**Comment by AnalogWorld** [comment]
Thanks All I am printing to a terminal.   I am going to investigate using RXD / TXD  lines on Pickit4 and Snap.   The nice thing about these tools is you can program  normal UPDI and hi voltage using MPLAB. (I wonder if AVRDUDE can send a command to do this?)  Also just recently they have added a plugin called MPLAB Data visualizer.  This seems like a powerful debugging tool however I am not sure about the communication through UPDI.  A lot of experimenting to do.

**Reply by SpenceKonde** [reply]
man, that sounds way more sophisticated than what I do.... this core was developed and tested with hacked serial adapters and a second serial adapter catching the data it sends back

**Reply by AnalogWorld** [reply]
Pretty sweet core and development board! Your documentation is is really good too. Its been a bit frustrating as if you want to use a ATTINY chip you can use Arduino , Studio 7 , MPLAB X IDE or one of the after market IDE's.  Not being a code guy Arduino is the most attractive solution with the most friendly learning curve but with limitations. Studio will not be supported any longer and MPLAB with a way more steep learning curve now has a plug in for Arduino (I Need to look at this).
So Megatinycore has been a really good tool . Please keep up the great work. Would like to see you on Youtube too (you could actually be monetized?)
![Serial print](https://github.com/SpenceKonde/megaTinyCore/assets/132411219/e7ba0ef8-e6af-424d-8589-4c3e17d3c420)
